File: apps/android/app/src/main/java/com/voiid/app/main/games/SnakeSkins.kt
package com.voiid.app.main.games

import androidx.compose.ui.graphics.Color

/**
 * Snake skin catalogue (docs/GAMES_SNAKE_VISUALS.md §2).
 *
 * THE SERVER SENDS AN ID, NOTHING ELSE. What a skin looks like lives here, so adding a
 * colourway is a client release and never a server one — and a client meeting an id it has
 * never heard of falls back to the plain palette colour rather than to nothing.
 *
 * MUST STAY IN SYNC WITH iOS `SnakeSkins.swift`. Band colours and band lengths are the values
 * that decide whether two players see the same arena, so they are duplicated deliberately and
 * identically rather than derived.
 */
data class SnakeSkin(
    /** Colours repeated along the body, head-first. */
    val bands: List<Color>,
    /** World units per band. Roughly one segment (14) unless a pattern wants finer stripes. */
    val bandLength: Float,
    /** Additive halo tint, or null for the body colour's own glow. */
    val glow: Color? = null,
    /** Face glyph drawn on the head, or null to keep the default eyes. */
    val face: String? = null,
)

object SnakeSkins {
    /** The launch set. Five are pure colour data; four carry a face. */
    private val catalogue: Map<String, SnakeSkin> = mapOf(
        "rainbow" to SnakeSkin(
            bands = listOf(
                Color(0xFFFF3B47), Color(0xFFFF8A2B), Color(0xFFFFD93D),
                Color(0xFF5CE65C), Color(0xFF4DA8FF), Color(0xFF5A5AF2),
                Color(0xFF9B5CFF),
            ),
            bandLength = 14f),

        "candy" to SnakeSkin(
            bands = listOf(Color.White, Color(0xFFFF70B8)),
            bandLength = 10f),

        "lava" to SnakeSkin(
            bands = listOf(Color(0xFFFF3B00), Color(0xFFFF8A2B), Color(0xFFFFD93D)),
            bandLength = 12f, glow = Color(0xFFFF731A)),

        "frost" to SnakeSkin(
            bands = listOf(Color(0xFF8DF7C8), Color(0xFF4DA8FF), Color.White),
            bandLength = 16f, glow = Color(0xFF66D9FF)),

        "shadow" to SnakeSkin(
            bands = listOf(Color(0xFF2B2B3F), Color(0xFF4A4A6A)),
            bandLength = 18f, glow = Color(0xFF7359BF)),

        "bunny" to SnakeSkin(
            bands = listOf(Color.White, Color(0xFFF5F5FF)),
            bandLength = 13f, face = "bunny"),

        "corgi" to SnakeSkin(
            bands = listOf(Color(0xFFE8A15C), Color(0xFFFFF1DC)),
            bandLength = 13f, face = "corgi"),

        "lion" to SnakeSkin(
            bands = listOf(Color(0xFFD9913F), Color(0xFFB4762E)),
            bandLength = 15f, face = "lion"),

        "unicorn" to SnakeSkin(
            bands = listOf(
                Color(0xFFFFBFD9), Color(0xFFD9CCFF),
                Color(0xFFBFF2FF), Color(0xFFFFF2BF),
            ),
            bandLength = 12f, glow = Color(0xFFFF73D9), face = "unicorn"),
    )

    /**
     * Resolve a skin id, falling back to a CHECKERED skin in the snake's palette colour.
     *
     * The fallback is what lets an old client meet a new skin without breaking — and it is
     * checkered rather than solid so an unknown skin still reads as a snake with segments
     * rather than as a featureless tube.
     */
    fun resolve(id: String?, fallback: Color): SnakeSkin =
        catalogue[id] ?: checkered(fallback)

    /**
     * A single colour turned into a box-box pattern.
     *
     * User testing asked for the checker "all ways through" even on a plain snake — a solid
     * body reads as a featureless tube and gives the eye nothing to judge length or speed
     * against. Alternating the colour with a darker shade of ITSELF keeps it recognisably one
     * colour (which is the point of picking it) while still segmenting.
     *
     * This is also what makes a custom picked colour work without needing its own catalogue
     * entry: any colour becomes a two-band skin.
     */
    fun checkered(base: Color): SnakeSkin = SnakeSkin(
        bands = listOf(base, shade(base, 0.72f)),
        bandLength = 13f,
    )

    private fun shade(c: Color, f: Float): Color =
        Color(red = c.red * f, green = c.green * f, blue = c.blue * f, alpha = c.alpha)
}

File: apps/ios/Voiid/Voiid/Games/SnakeSkins.swift
//
//  SnakeSkins.swift
//  Voiid
//
//  Snake skin catalogue (docs/GAMES_SNAKE_VISUALS.md §2).
//
//  THE SERVER SENDS AN ID, NOTHING ELSE. What a skin looks like lives here, so adding a
//  colourway is a client release and never a server one — and a client meeting an id it has
//  never heard of falls back to the plain palette colour rather than to nothing.
//
//  MUST STAY IN SYNC WITH Android `SnakeSkins.kt`. Band colours and band lengths are the
//  values that decide whether two players see the same arena, so they are duplicated
//  deliberately and identically rather than derived.
//

import simd

/// One skin: how a body is banded, and what sits on the head.
struct SnakeSkin {
    /// Colours repeated along the body, head-first.
    let bands: [SIMD4<Float>]
    /// World units per band. Roughly one segment (14) unless a pattern wants finer stripes.
    let bandLength: Float
    /// Additive halo tint, or nil for the body colour's own glow.
    let glow: SIMD4<Float>?
    /// Face glyph drawn on the head, or nil to keep the default eyes.
    let face: String?

    static func rgb(_ r: Double, _ g: Double, _ b: Double) -> SIMD4<Float> {
        SIMD4(Float(r), Float(g), Float(b), 1)
    }
}

enum SnakeSkins {
    /// The launch set. Five are pure colour data; four carry a face.
    static let catalogue: [String: SnakeSkin] = [
        "rainbow": SnakeSkin(
            bands: [
                rgb(1.00, 0.23, 0.28), rgb(1.00, 0.54, 0.17), rgb(1.00, 0.85, 0.24),
                rgb(0.36, 0.90, 0.36), rgb(0.30, 0.66, 1.00), rgb(0.35, 0.35, 0.95),
                rgb(0.61, 0.36, 1.00),
            ],
            bandLength: 14, glow: nil, face: nil),

        "candy": SnakeSkin(
            bands: [rgb(1, 1, 1), rgb(1.00, 0.44, 0.72)],
            bandLength: 10, glow: nil, face: nil),

        "lava": SnakeSkin(
            bands: [rgb(1.00, 0.23, 0.00), rgb(1.00, 0.54, 0.17), rgb(1.00, 0.85, 0.24)],
            bandLength: 12, glow: rgb(1.00, 0.45, 0.10), face: nil),

        "frost": SnakeSkin(
            bands: [rgb(0.55, 0.97, 0.78), rgb(0.30, 0.66, 1.00), rgb(1, 1, 1)],
            bandLength: 16, glow: rgb(0.40, 0.85, 1.00), face: nil),

        "shadow": SnakeSkin(
            bands: [rgb(0.17, 0.17, 0.25), rgb(0.29, 0.29, 0.42)],
            bandLength: 18, glow: rgb(0.45, 0.35, 0.75), face: nil),

        "bunny": SnakeSkin(
            bands: [rgb(1, 1, 1), rgb(0.96, 0.96, 1.00)],
            bandLength: 13, glow: nil, face: "bunny"),

        "corgi": SnakeSkin(
            bands: [rgb(0.91, 0.63, 0.36), rgb(1.00, 0.95, 0.86)],
            bandLength: 13, glow: nil, face: "corgi"),

        "lion": SnakeSkin(
            bands: [rgb(0.85, 0.57, 0.25), rgb(0.71, 0.46, 0.18)],
            bandLength: 15, glow: nil, face: "lion"),

        "unicorn": SnakeSkin(
            bands: [
                rgb(1.00, 0.75, 0.85), rgb(0.85, 0.80, 1.00),
                rgb(0.75, 0.95, 1.00), rgb(1.00, 0.95, 0.75),
            ],
            bandLength: 12, glow: rgb(1.00, 0.45, 0.85), face: "unicorn"),
    ]

    private static func rgb(_ r: Double, _ g: Double, _ b: Double) -> SIMD4<Float> {
        SnakeSkin.rgb(r, g, b)
    }

    /// Resolve a skin id, falling back to a CHECKERED skin in the snake's palette colour.
    ///
    /// The fallback is what lets an old client meet a new skin without breaking — and it is
    /// checkered rather than solid so an unknown skin still reads as a snake with segments
    /// rather than as a featureless tube.
    static func resolve(_ id: String?, fallback: SIMD4<Float>) -> SnakeSkin {
        if let id, let skin = catalogue[id] { return skin }
        // `custom:RRGGBB` from the colour picker. Rendered as a checkered two-band skin like
        // any other, which is why a custom colour needs no catalogue entry at all.
        if let id, id.hasPrefix("custom:"), let rgb = UInt32(id.dropFirst(7), radix: 16) {
            return checkered(SIMD4(
                Float((rgb >> 16) & 0xFF) / 255,
                Float((rgb >> 8) & 0xFF) / 255,
                Float(rgb & 0xFF) / 255,
                1))
        }
        return checkered(fallback)
    }

    /// A single colour turned into a box-box pattern.
    ///
    /// User testing asked for the checker "all ways through" even on a plain snake — a solid
    /// body reads as a featureless tube and gives the eye nothing to judge length or speed
    /// against. Alternating the colour with a darker shade of ITSELF keeps it recognisably
    /// one colour (which is the point of picking it) while still segmenting, and it is what
    /// lets a custom picked colour work without its own catalogue entry.
    static func checkered(_ base: SIMD4<Float>) -> SnakeSkin {
        SnakeSkin(
            bands: [base, SIMD4(base.x * 0.72, base.y * 0.72, base.z * 0.72, base.w)],
            bandLength: 13, glow: nil, face: nil)
    }
}
